packages/wallet-types/src/types.ts

```diff
@@ -1,8 +1,114 @@
+import {PublicKey} from "./public-keys";
+import {NetworkConfig} from "./network-info";
+import {OfflineSigner, TxsResponse} from "@cosmjs/launchpad";
+import {OfflineDirectSigner} from "@cosmjs/proto-signing";
+
+export interface Account {
+    address: string
+    publicKey: PublicKey
+}
+
+export interface Signature {
+    readonly publicKey: PublicKey
+    readonly signature: Uint8Array;
+}
+
+export interface NetworksApi {
+    currentNetwork(): Promise<NetworkConfig | undefined>;
+    switchToNetwork(network: NetworkConfig): Promise<void>;
+    switchToNetworkByChainId(chainId: string): Promise<void>;
+    listNetworks(): Promise<NetworkConfig[]>;
+}
+
+export interface AccountsApi {
+    switchAccount(address: string): Promise<void>;
+    listAccounts(): Promise<Account[]>;
+    getAccount(chainId: string): Promise<Account>;
+}
+
+export interface AddressBookEntry {
+    address: string;
+    name: string;
+    chainIds: string[];
+}
+
+export interface AddressBookApi {
+    listEntries(): Promise<AddressBookEntry[]>;
+    addEntry(entry: AddressBookEntry): Promise<void>;
+    updateEntry(entry: AddressBookEntry): Promise<void>;
+    deleteEntry(address: string): Promise<void>;
+}
+
+export interface EventHandler<T> {
+    subscribe(handler: T): void;
+    unsubscribe(handler: T): void;
+}
+
+export interface EventsApi {
+    onStatusChanged: EventHandler<(status: WalletStatus) => void | Promise<void>>;
+    onNetworkChanged: EventHandler<(network: NetworkConfig) => void | Promise<void>>;
+    onAccountChanged: EventHandler<(account: Account) => void | Promise<void>>;
+    onTxSuccessful: EventHandler<(tx: TxResponse) => void | Promise<void>>;
+    onTxFailed: EventHandler<(tx: TxsResponse) => void | Promise<void>>;
+}
+
+export interface SigningApi {
+    signArbitrary(
+        chainId: string,
+        signer: string,
+        data: string | Uint8Array
+    ): Promise<Signature>;
+
+    verifyArbitrary(
+        chainId: string,
+        signer: string,
+        data: string | Uint8Array,
+        signature: Signature
+    ): Promise<boolean>;
+
+    getOfflineSigner(chainId: string): Promise<OfflineSigner | OfflineDirectSigner>;
+    getOfflineDirectSigner(chainId: string): Promise<OfflineDirectSigner>;
+    getOfflineAminoSigner(chainId: string): Promise<OfflineSigner>;
+}
+
+export type WalletStatus = "locked" | "unlocked";
+
+export interface WalletApi {
+
+    /**
+     * WALLET STATUS API
+     */
+
+    /**
+     * Determine the status of the wallet. It is either locked or unlocked
+     */
+    status(): Promise<WalletStatus>;
+
+    /**
+     * Allows the user to lock the wallet from the UI. This can implement logout like functionality
+     */
+    lockWallet(): Promise<void>;
+
+    /**
+     * Allows the dApp to
+     */
+    unlockWallet(): Promise<void>;
+
+    /**
+     * NETWORK SELECTION API
+     */
+    networks: NetworksApi;
+    accounts: AccountsApi;
+    addressBook: AddressBookApi;
+    signing: SigningApi;
+}
+
```

Comment:
We can add a message/notification api. It should be on a per account/publicKey basis as messages can be only fetched for accounts that are authenticated via the auth module (1 public key at one time).

Reply:
Also accountsAPI can store assets per chainId by extending the currency type to include balances/decimals. Also NFTs?

Reply:
My thoughts:

* **Messaging API** - I think that we need to think a little bit about this one before opening it up via the browser API. My initial thinking is that it probably would look a bit more like a REST interface that we run centrally. At least for the first version. There are lots of things to consider here specifically things like Abuse which we would want to have control on first. **TL;DR** I don't think we add it in this version but there the API is structured so it can be introduced in the future in a backwards compatible way
* **NFT API** - Yes I agree, we should add this. I think that it probably comes in with v0.2 though. It is not 100% clear to me the sort of API we would want for this now. If it is clear to you then would be happy to add.